chronos/asyncfutures2.nim

```diff
@@ -784,21 +799,83 @@ proc cancelAndWait*(fut: FutureBase): Future[void] =
   ##
   ## If ``fut`` is already finished (completed, failed or cancelled) result
   ## Future[void] object will be returned complete.
-  var retFuture = newFuture[void]("chronos.cancelAndWait(T)")
-  proc continuation(udata: pointer) =
-    if not(retFuture.finished()):
-      retFuture.complete()
-  proc cancellation(udata: pointer) =
-    if not(fut.finished()):
-      fut.removeCallback(continuation)
+  let retFuture = newFuture[void]("chronos.cancelAndWait(FutureBase)",
+                                  {FutureFlag.OwnCancelSchedule})
+
+  proc continuation(udata: pointer) {.gcsafe.} =
+    retFuture.complete()
+
+  proc cancellation(udata: pointer) {.gcsafe.} =
+    # We are not going to change the state of `retFuture` to cancelled, so we
+    # will prevent the entire sequence of Futures from being cancelled one more
+    # time.
+    discard
+
   if fut.finished():
     retFuture.complete()
   else:
     fut.addCallback(continuation)
     retFuture.cancelCallback = cancellation
     # Initiate cancellation process.
     fut.cancel()
-  return retFuture
+  retFuture
+
+proc checkedCancelAndWait*(fut: FutureBase): Future[bool] =
```

Comment:
it's unclear what this function does compared to `cancelAndWait` - when should one be used and when should the other be used?

Reply:
removed.